core/examples/http_bench.rs

```diff
@@ -183,178 +174,121 @@ pub fn bad_resource() -> Error {
   Error::new(ErrorKind::NotFound, "bad resource id")
 }
 
-struct TcpListener(tokio::net::TcpListener);
-
-impl Resource for TcpListener {}
-
-struct TcpStream(tokio::net::TcpStream);
-
-impl Resource for TcpStream {}
-
-lazy_static! {
-  static ref RESOURCE_TABLE: Mutex<ResourceTable> =
-    Mutex::new(ResourceTable::default());
+#[derive(Default)]
+pub struct HttpBenchStateInner {
+  resource_table: ResourceTable,
 }
 
-fn lock_resource_table<'a>() -> MutexGuard<'a, ResourceTable> {
-  RESOURCE_TABLE.lock().unwrap()
-}
+#[derive(Clone, Default)]
+pub struct HttpBenchState(Rc<RefCell<HttpBenchStateInner>>);
 
-struct Accept {
-  rid: ResourceId,
+impl HttpBenchState {
+  pub fn new() -> Self {
+    Default::default()
+  }
 }
 
-impl Future for Accept {
-  type Output = Result<(tokio::net::TcpStream, SocketAddr), std::io::Error>;
-
-  fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
-    let inner = self.get_mut();
-
-    let mut table = lock_resource_table();
-    match table.get_mut::<TcpListener>(inner.rid) {
-      None => Poll::Ready(Err(bad_resource())),
-      Some(listener) => {
-        let listener = &mut listener.0;
-        listener.poll_accept(cx)
-      }
-    }
+impl Deref for HttpBenchState {
+  type Target = RefCell<HttpBenchStateInner>;
+  fn deref(&self) -> &Self::Target {
+    &*self.0
   }
 }
 
 fn op_accept(
+  state: HttpBenchState,
   record: Record,
-  _zero_copy_buf: Option<ZeroCopyBuf>,
+  _buf: Option<ZeroCopyBuf>,
 ) -> Pin<Box<HttpOp>> {
   let rid = record.arg as u32;
-  debug!("accept {}", rid);
-
-  let fut = async move {
-    let (stream, addr) = Accept { rid }.await?;
-    debug!("accept success {}", addr);
-    let mut table = lock_resource_table();
-    let rid = table.add("tcpStream", Box::new(TcpStream(stream)));
-    Ok(rid as i32)
-  };
-
-  fut.boxed()
+  debug!("accept rid={}", rid);
+
+  poll_fn(move |cx| {
+    let resource_table = &mut state.borrow_mut().resource_table;
+    let listener = resource_table
+      .get_mut::<TcpListener>(rid)
+      .ok_or_else(bad_resource)?;
+    listener.poll_accept(cx).map_ok(|(stream, _addr)| {
+      let rid = resource_table.add("tcpStream", Box::new(stream));
+      rid as i32
+    })
+  })
```

Comment:
much nicer

Reply:
Since there is no mutex on resource table, can we get rid of `poll_fn()` entirely?
```rust
let resource_table = &mut state.borrow_mut().resource_table;
let listener = resource_table
    .get_mut::<TcpListener>(rid)
    .ok_or_else(bad_resource)?;
let (stream, _addr) = listener.accept().await?;
resource_table.add("tcpStream", Box::new(stream))
```

Reply:
Nvm, that would lead to multiple mutable borrows which is obviosuly not permitted.

Reply:
> Nvm, that would lead to multiple mutable borrows which is obviosuly not permitted.

Indeed. That's why only listen() and close() can use an `async {}` block.